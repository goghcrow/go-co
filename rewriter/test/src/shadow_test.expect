// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	. "github.com/goghcrow/go-co"
	ʂɘʠ "github.com/goghcrow/go-co/seq"
)

func shadow0() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := 0
				return ʂɘʠ.For[int](func() bool {
					return i < 2
				}, func() {
					i++
				}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Normal[int]()
					})
				}))
			})
		}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

			i++
			println(i)
			return ʂɘʠ.Return[int]()
		}))
	}))

}

// ↑ is equivalent to ↓
func shadow1() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := 0
				return ʂɘʠ.For[int](func() bool {
					return i < 2
				}, func() {
					i++
				}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Normal[int]()
					})
				}))
			})
		}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

			i++
			println(i)
			return ʂɘʠ.Return[int]()
		}))
	}))

}
