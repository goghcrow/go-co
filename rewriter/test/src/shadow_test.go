package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
)

// 1. Init stmt in ForStmt will up to outer scope
// 2. Key, Value Expr in RangeStmt will be down to body scope

func forInit() Iter[int] {
	i := 0
	for i := 0; i < 3; i++ {
		i := i + 1
		Yield(i)
	}
	Yield(i)
	return nil
}

// ↑ is equivalent to ↓
func forInit1() Iter[int] {
	i := 0
	{
		i := 0
		for ; i < 3; i++ {
			i := i + 1
			Yield(i)
		}
	}
	Yield(i)
	return nil
}

func rangeInit() Iter[int] {
	i := 0
	for _, i := range []int{0, 1, 2} {
		i := i + 1
		Yield(i)
	}
	Yield(i)
	return nil
}

// ↑ is equivalent to ↓
func rangeInit1() Iter[int] {
	i := 0
	for _, i := range []int{0, 1, 2} {
		{
			i := i + 1
			Yield(i)
		}
	}
	Yield(i)
	return nil
}

func TestShadow(t *testing.T) {
	assertEqual(t, iter2slice(forInit()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(forInit1()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(rangeInit()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(rangeInit1()), []int{1, 2, 3, 0})
}
