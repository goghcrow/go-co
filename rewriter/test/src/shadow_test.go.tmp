//go:build !co
// +build !co

// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	"testing"

	. "github.com/goghcrow/go-co"
	ʂɘʠ "github.com/goghcrow/go-co/seq"
)

func forInit() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := 0
				return ʂɘʠ.For[int](func() bool {
					return i < 3
				}, func() {
					i++
				}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

					i := i + 1
					return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Normal[int]()
					})
				}))
			})
		}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Return[int]()
			})
		}))
	}))

}

// ↑ is equivalent to ↓
func forInit1() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := 0
				return ʂɘʠ.For[int](func() bool {
					return i < 3
				}, func() {
					i++
				}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

					i := i + 1
					return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Normal[int]()
					})
				}))
			})
		}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Return[int]()
			})
		}))
	}))

}

func rangeInit() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		ɪʇ := ʂɘʠ.NewSliceIter([]int{0, 1, 2})
		return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.While[int](func() bool {
				return ɪʇ.MoveNext()
			}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := ɪʇ.Current().Val
				return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

					i := i + 1
					return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
						return ʂɘʠ.Normal[int]()
					})
				})
			}))
		}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Return[int]()
			})
		}))
	}))

}

// ↑ is equivalent to ↓
func rangeInit1() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		ɪʇ := ʂɘʠ.NewSliceIter([]int{0, 1, 2})
		return ʂɘʠ.Combine[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.While[int](func() bool {
				return ɪʇ.MoveNext()
			}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := ɪʇ.Current().Val
				return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
					return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

						i := i + 1
						return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
							return ʂɘʠ.Normal[int]()
						})
					})
				})
			}))
		}), ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
			return ʂɘʠ.Bind[int](i, func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Return[int]()
			})
		}))
	}))

}

func TestShadow(t *testing.T) {
	assertEqual(t, iter2slice(forInit()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(forInit1()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(rangeInit()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(rangeInit1()), []int{1, 2, 3, 0})
}
