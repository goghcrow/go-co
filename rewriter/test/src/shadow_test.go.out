//go:build !co

// Code generated by github.com/goghcrow/go-co DO NOT EDIT.
package src

import (
	ʂɘʠ "github.com/goghcrow/go-co/seq"
	"testing"
)

func forInit() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		return ʂɘʠ.Combine[int](
			ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := 0
				return ʂɘʠ.For[int](func() bool {
					return i < 3
				}, func() {
					i++
				}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

					i := i + 1
					return ʂɘʠ.Bind[int](i,
						ʂɘʠ.Normal[int],
					)
				}))
			}),
			ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Bind[int](i,
					ʂɘʠ.Return[int],
				)
			}))
	}))

}

// ↑ is equivalent to ↓
func forInit1() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		return ʂɘʠ.Combine[int](
			ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

				i := 0
				return ʂɘʠ.For[int](func() bool {
					return i < 3
				}, func() {
					i++
				}, ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

					i := i + 1
					return ʂɘʠ.Bind[int](i,
						ʂɘʠ.Normal[int],
					)
				}))
			}),
			ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Bind[int](i,
					ʂɘʠ.Return[int],
				)
			}))
	}))

}

func rangeInit() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		ɪʇ := ʂɘʠ.NewSliceIter([]int{0, 1, 2})
		return ʂɘʠ.Combine[int](
			ʂɘʠ.While[int](
				ɪʇ.MoveNext,
				ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

					i := ɪʇ.Current().Val
					return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

						i := i + 1
						return ʂɘʠ.Bind[int](i,
							ʂɘʠ.Normal[int],
						)
					})
				})),
			ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Bind[int](i,
					ʂɘʠ.Return[int],
				)
			}))
	}))

}

// ↑ is equivalent to ↓
func rangeInit1() ʂɘʠ.Iterator[int] {
	return ʂɘʠ.Start[int](ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
		i := 0
		ɪʇ := ʂɘʠ.NewSliceIter([]int{0, 1, 2})
		return ʂɘʠ.Combine[int](
			ʂɘʠ.While[int](
				ɪʇ.MoveNext,
				ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

					i := ɪʇ.Current().Val
					return ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {

						i := i + 1
						return ʂɘʠ.Bind[int](i,
							ʂɘʠ.Normal[int],
						)
					})

				})),
			ʂɘʠ.Delay[int](func() ʂɘʠ.Seq[int] {
				return ʂɘʠ.Bind[int](i,
					ʂɘʠ.Return[int],
				)
			}))
	}))

}

func TestShadow(t *testing.T) {
	assertEqual(t, iter2slice(forInit()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(forInit1()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(rangeInit()), []int{1, 2, 3, 0})
	assertEqual(t, iter2slice(rangeInit1()), []int{1, 2, 3, 0})
}
